### ISSUE TITLE
[question] an alternative for binary cross entropy

### ISSUE THREAD

**Issue by yasserzamani** [zackchase/mxnet-the-straight-dope#493]
In Chapter 2, logistic regression, I played with your codes to see why it works. I constructed following loss function instead:
```python
def log_loss(output, y):
    yhat = logistic(output)
    return  -nd.log(1 - nd.abs(yhat - y))
```
based on my following reasoning:

* firstly I was interested in distance between my prediction and actual value i.e. `nd.abs(yhat-y)`
* then, as they are small values between 0 and 1, to amplify the error I added `nd.log`
* then, to assign nearest values to 0 and further values to Infinity I added `1 - `
* then, to map to (+0, +Infinity) I added `-`

Surprisingly I got almost same accuracy as your binary cross entropy! I restart kernel and ran it several times and always I got an accuracy near to `0.83`.

Now my question is simple, why cross entropy is famous and my above constructed isn't? :) I couldn't understand *why and how* cross entropy loss function is better.

**Comment by zackchase** [comment]
Well you can get away with quite a lot of arbitrary nosense as long as you don't mess up the *sign* of the gradient. In this case, that's the one thing that's not messed up.

You ask what is so special about cross entropy. The term *cross entropy* might be filing you in to thinking something fancy is going on here. But really in the special case here where y (the true label) is lways a onehot vector, cross entropy is simply saying maximize the log likelihood assigned to the correct label. That's it.

So that makes cross entropy loss a proper loss function and we can interpret the outputs of our classifier as estimated probabilities. However, for your loss function who knows what comes out.

**Comment by yasserzamani** [comment]
Interestingly this time I thought to penalize more (i.e. bigger error) when Euclidean distance is more and penalize less vice versa. So I thought `nd.tan` is a nice option as following loss function:
```python
def log_loss(output, y):
    yhat = logistic(output)
    return nd.tan(nd.arcsin(nd.abs(yhat - y))*9999/10000)
```
* Firstly I mapped Euclidean distance to angle by `nd.arcsin`
* Then `nd.tan` penalizes more when more angle
* Then `*9999/10000` is there to avoid NaN

Interestingly this time I saw this loss function continues good decreasing even after 30 epochs, so, I increased your epochs to 90 where it is going to being stable. I always get `.84` accuracy there :smiley: 

I thought *what makes a loss function better than other?* and practiced following results:
* using only Euclidean distance itself: `.75` accuracy and no significant improvements increasing epochs
* using (1 / Euclidean distance itself): same as above
* using log(Euclidean distance): `.83` accuracy and no significant improvements increasing epochs
* using tan(Euclidean distance): `.84` accuracy with significant improvements increasing epochs till 90

VERY INTERESTING! and I know there are a lot of functions in math which I didn't practiced!

I still cannot visualize in my mind how cross entropy and these functions have such results? I mean why using only Euclidean distance itself doesn't work well here (while it works well in previous chapter)? why using tan(Euclidean distance) has same accuracy? why log(Euclidean distance) cannot have better accuracy than tan? and I see binary cross entropy currently is in first place with `.844` accuracy in my several runs, why? I'm trying to discover these `why`s because knowing answers to them, learns me how to define loss function i.e. when to use log, when tan, when exp, when 1/, when L2 and etc :smiley:
